text/0000-di-hoc.md

```diff
@@ -0,0 +1,78 @@
+* Start Date: 2018-09-05
+* RFC PR: (leave this empty)
+* Fusion Issue: (leave this empty)
+
+# Summary
+
+A dependency-injection higher-order component for consuming plugins in container components for FusionJS.
+
+# Basic example
+
+The DI HOC API would mirror [middleware DI](https://fusionjs.com/api/fusion-core#middleware) for components.
+
+```jsx
+import React from 'react';
+import {withPlugins} from 'fusion-react';
+import {LoggerToken} from 'fusion-tokens';
+
+export function Foo({ logger }) {
+  return (
+    <button onclick={() => logger.info('clicked')}>
+      Click me
+    </button>
+  );
+}
+
+export default withPlugins(
+  { logger: LoggerToken },
+  Foo
+);
+```
+
+# Motivation
+
+The current pattern for consuming services from the dependency-injection system relies on plugin authors to expose their own higher-order components (ex. [i18n plugin HOC](https://github.com/fusionjs/fusion-plugin-i18n-react/blob/master/src/hoc.js)). The `fusion-react` module makes this fairly easy with the `ProvidedHOC` creator, but it still requires work on the part of plugin authors and may require that a single container component be wrapped with multiple HOCs. This gets hard to read and manage.
+
+Additionally, accessing the HOCs requires component implenentors to import the HOC from a specific package which breaks the spirit of DI's decoupling and may lead to component-consumer confusion as the API they recieve from the HOC may not be the same as the one provided by the module from which they imported the HOC.
+
+I propose that `fusion-react` should provide its own DI HOC with the same API as middleware. With this pattern, container component implementors need only rely on fusion itself and tokens. Plugins would no longer need to provide their own HOCs, only one HOC is required per container, and components can be decoupled from plugin implementations.
+
+# Detailed design
+
+The DI HOC should use the same or similar token resolution as middleware DI such that these are equivalent:
+
+```js
+app.middleware({m3: M3Token}, ({m3}) => {
+  // etc
+});
+```
+
+```jsx
+withPlugins({m3: M3Token}, ({m3}) => {
+  return (
+    <div>
+      // etc
+    </div>
+  );
+});
+```
+
+# Drawbacks
+
+In some cases, the HOCs created by plugins may intentionally alter the API exposed to React components. Defering to a generic DI HOC may require plugins to change their internals to remain compatible.
```

Comment:
perhaps we could introduce an RFC in `fusion-core` that adds an optional argument to `createPlugin`'s argument map that defines the behavior of exposing a plugin via this proposed HOC

something like `expose` where `false` disables exposing via the HOC, `true` naively exposes it, and a function defines any setup behavior and returns the service, that could both help to alleviate this point and add more utility and "safety" to DI

Reply:
Perhaps it would be better to allow a `mapServicesToProps` callback argument similar to `react-redux`'s `connect` function.